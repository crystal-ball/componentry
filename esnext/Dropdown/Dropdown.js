import withActive from '../withActive/withActive';
import activeContainer from '../component-factories/active-container';
import activeContent from '../component-factories/active-content';
import activeTrigger from '../component-factories/active-trigger';
const Content = activeContent({
  componentArias: {
    labelledby: true,
    hidden: true
  },
  element: 'dropdown',
  name: 'DropdownContent'
});
const withActiveContent = withActive()(Content);
const Trigger = activeTrigger({
  element: 'dropdown',
  componentArias: {
    expanded: true,
    haspopup: true,
    id: true
  },
  name: 'DropdownTrigger'
});
const withActiveTrigger = withActive()(Trigger);
const Item = activeTrigger({
  // Suppress the default btn classes, the .dropdown-item classes have all styles
  // needed for button items
  baseButton: true,
  classes: 'dropdown-item',
  componentArias: {},
  name: 'DropdownItem'
});
const DropdownItem = withActive()(Item);
const Dropdown = activeContainer({
  Content: withActiveContent,
  Trigger: withActiveTrigger,
  defaultDirection: 'bottom',
  element: 'dropdown',
  escHandler: true,
  externalClickHandler: true,
  name: 'Dropdown'
});
Dropdown.Content = withActiveContent;
Dropdown.Trigger = withActiveTrigger;
Dropdown.Item = DropdownItem;
/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */

export default Dropdown;