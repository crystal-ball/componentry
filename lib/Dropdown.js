/* eslint-disable react/prop-types */
import { Component } from 'react';

import toggleable from './HOCs/toggleableHOC';
import ToggleContent from './ToggleContent';
import ToggleTrigger from './ToggleTrigger';
import { closest } from './utils/dom';

/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Menu Items with class names 'dropdown-item'
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */
class Dropdown extends Component {
  constructor(props) {
    super(props);
    this.clickHandler = this.clickHandler.bind(this);
    this.keyHandler = this.keyHandler.bind(this);
    this.toggleActive = this.toggleActive.bind(this);
  }

  clickHandler(e) {
    // If the click was ouside dropdown, close the dropdown and then cleanup the listener
    if (!closest(e.target, `${this.guid}-container`)) {
      this.toggleActive();
    }
  }

  keyHandler(e) {
    // Escape key is which 27, when escape key is hit, toggle state
    if (e.which === 27) {
      this.toggleActive();
    }
  }

  toggleActive(e) {
    const { onActivate, onActivated, onDeactivate, onDeactivated } = this.props;
    let { active } = this.props;
    if (active === undefined) {
      active = this.state.active;
    }

    if (!active) {
      // If the dropdown is closed, it's now opening, so setup event listeners
      onActivate(this, e); // Consumer hooks
      document.addEventListener('mouseup', this.clickHandler);
      document.addEventListener('touchend', this.clickHandler);
      document.addEventListener('keydown', this.keyHandler);
    } else {
      // If the dropdown is open, it's now closing, so remove event listeners
      onDeactivate(this, e); // Consumer hooks
      document.removeEventListener('mouseup', this.clickHandler);
      document.removeEventListener('touchend', this.clickHandler);
      document.removeEventListener('keydown', this.keyHandler);
    }

    if (this.props.active === undefined) {
      // State is not controlled, update internal state
      this.setState({ active: !active });
    }

    if (!active) {
      // If the dropdown is closed, it's now opening, so setup event listeners
      onActivated(this, e); // Consumer hooks
    } else {
      // If the dropdown is open, it's now closing, so remove event listeners
      onDeactivated(this, e); // Consumer hooks
    }
  }
}

Dropdown.Content = ToggleContent;
Dropdown.Trigger = ToggleTrigger;

const toggleableDropdown = toggleable({
  contentArias: { labelledby: true, hidden: true },
  elementType: 'dropdown',
  triggerArias: { expanded: true, haspopup: true, id: true }
})(Dropdown);

export default toggleableDropdown;
