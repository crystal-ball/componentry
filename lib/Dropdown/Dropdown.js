import {
  StateContainer,
  contentFactory,
  triggerFactory,
  withActive,
  withState
} from '../State'

/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Menu Items with class names 'dropdown-item'
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */
export default withState({ type: 'dropdown' })(
  class Drawer extends StateContainer {
    static Content = withActive({ labelledby: true, hidden: true })(
      contentFactory()
    )
    static Toggle = withActive({ expanded: true, haspopup: true, id: true })(
      triggerFactory({ link: false })
    )
  }
)
