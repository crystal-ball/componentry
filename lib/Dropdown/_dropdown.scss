// ========================================================
// Dropdowns
// ========================================================

// BS dropdowns expect a containing .dropdown and .dropdown-menu class, but for
// simplicity we use a wrapping `<TYPE>-container` and `<TYPE>-content` on all
// elements, and extend the original dropdown here
.dropdown-container {
  @extend .dropdown;
}
.dropdown-content {
  @extend .dropdown-menu;
}

.dropdown-content {
  // In Bs this is applied by adding .show class
  display: block;

  // This is the only theme color style not automatically suppressed by BS
  .dropdown-item {
    box-shadow: none;
  }
}
