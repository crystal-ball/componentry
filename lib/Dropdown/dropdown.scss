// ========================================================
// Dropdowns
// ========================================================

.dropdown {
  // Defaulting to inline-block so that click area is constrained
  // to dropdown button area.
  display: inline-block;
}

// In Bs show/hide is handled by adding a .show class, Componentry uses aria-hidden
// to hide content
.dropdown-content {
  display: block;
}

// Add styles for inline dropdowns for a border bottom that matches the anchor color
// Helps the dropdown to stand out, even when used inline style
.dropdown-toggle.btn-anchor {
  border-bottom: 1px solid $dropdown-trigger-border-color;

  @include hover-focus {
    border-bottom: 1px solid $dropdown-trigger-hover-border-color;
    text-decoration: none; // Trigger border is the decoration
  }
}
