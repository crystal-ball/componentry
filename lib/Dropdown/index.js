// @flow
import withState from '../HOCs/withState'
import withActive from '../HOCs/withActive'
import Button from '../Button'
import elementFactory from '../component-factories/element-factory'
import stateContainer from '../component-factories/state-container'
import stateContent from '../component-factories/state-content'
import stateTrigger from '../component-factories/state-trigger'

const Dropdown = withState(stateContainer({ element: 'dropdown' }))

Dropdown.Content = withActive(
  stateContent({
    element: 'dropdown',
    componentArias: { labelledby: true, hidden: true }
  })
)

Dropdown.Trigger = withActive(
  stateTrigger({
    element: 'dropdown',
    link: false,
    componentArias: { expanded: true, haspopup: true, id: true }
  })
)

Dropdown.Item = elementFactory({
  classes: 'dropdown-item',
  name: 'DropdownItem',
  tag: Button
})

/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */
export default Dropdown
