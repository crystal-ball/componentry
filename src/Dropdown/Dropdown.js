import activeContainer from '../active-container-factory'
import activeContent from '../active-content-factory'
import activeTrigger from '../active-trigger-factory'
import withActive from '../withActive'
import withTheme from '../withTheme'

const Content = activeContent({
  arias: { labelledby: true, hidden: true },
  classes: 'dropdown-content',
})

const Trigger = activeTrigger({
  arias: { expanded: true, haspopup: true, id: true },
  classes: 'dropdown-toggle',
})

const Item = activeTrigger({
  // TODO: what arias should this have?
  arias: {},
  classes: 'dropdown-item',
  // Dropdown items have appropriate styles
  defaultLink: false,
})

const Dropdown = activeContainer({
  Content: withTheme('DropdownContent', withActive(Content)),
  Trigger: withTheme('DropdownTrigger', withActive(Trigger)),
  classes: 'dropdown',
  escHandler: true,
  clickHandler: true,
})
Dropdown.Item = withTheme('DropdownItem', withActive(Item))
Dropdown.defaultProps = {
  ...Dropdown.defaultProps,
  direction: 'bottom',
}

/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */
export default withTheme('Dropdown', Dropdown)
