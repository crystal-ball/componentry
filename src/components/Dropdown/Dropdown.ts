import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../../utils/types'
import { Button } from '../Button/Button'

export interface DropdownProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface DropdownActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownItemProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'unstyled'
}

export interface Dropdown {
  (props: DropdownProps): React.ReactElement
  /**
   * [Dropdown action component üìù](https://componentry.design/components/dropdown)
   */
  Action: React.FC<DropdownActionProps>
  /**
   * [Dropdown content component üìù](https://componentry.design/components/dropdown)
   */
  Content: React.FC<DropdownContentProps>
  /**
   * [Dropdown item component üìù](https://componentry.design/components/dropdown)
   */
  Item: React.FC<DropdownItemProps>
}

/**
 * [Dropdown component üìù](https://componentry.design/components/dropdown)
 * @experimental
 */
export const Dropdown = activeContainerBuilder<DropdownProps>('Dropdown', {
  clickEvents: true,
  direction: 'bottom',
  escEvents: true,
}) as Dropdown

Dropdown.Action = activeActionBuilder<DropdownActionProps>('DropdownAction', {
  aria: { expanded: true, haspopup: true, id: true },
  defaultAs: Button,
})

Dropdown.Content = activeContentBuilder<DropdownContentProps>('DropdownContent', {
  aria: { labelledby: true, hidden: true },
})

Dropdown.Item = activeActionBuilder<DropdownItemProps>('DropdownItem')
