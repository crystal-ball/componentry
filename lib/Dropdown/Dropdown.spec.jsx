import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import Dropdown from './index'

describe('<Dropdown />', () => {
  test('should render appropriate classes and arias by default', () => {
    const wrapper = mount(
      <Dropdown>
        <Dropdown.Trigger>Trigger</Dropdown.Trigger>
        <Dropdown.Content>
          <Dropdown.Item>Item 1</Dropdown.Item>
          <Dropdown.Item>Item 2</Dropdown.Item>
        </Dropdown.Content>
      </Dropdown>
    )

    // Container
    expect(wrapper.find('.dropdown').length).toEqual(1)
    // Trigger
    const trigger = wrapper.find('button.dropdown-toggle')
    const guid = trigger.prop('id')
    expect(trigger.length).toEqual(1)
    expect(trigger.text()).toEqual('Trigger')
    expect(trigger.prop('aria-haspopup')).toEqual('true')
    expect(trigger.prop('aria-expanded')).toEqual('false')
    // Content
    // (labelledby should match the id of the trigger element)
    const content = wrapper.find('.dropdown-content')
    expect(content.length).toEqual(1)
    expect(content.text()).toEqual('Item 1Item 2')
    expect(content.prop('aria-labelledby')).toEqual(guid)
    expect(content.prop('aria-hidden')).toEqual('true')
    expect(wrapper.find('button.dropdown-item').length).toEqual(2)

    // Simulate a click, arias should update
    trigger.simulate('click')

    // Cannot re-use found nodes, the references are stale
    expect(wrapper.find('.btn.dropdown-toggle').prop('aria-expanded')).toEqual(
      'true'
    )
    // expect(wrapper.find('.dropdown-toggle').length).toEqual(1)
    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('false')
  })

  test('should call user event hooks', () => {
    const onActivate = jest.fn()
    const onActivated = jest.fn()
    const onDeactivate = jest.fn()
    const onDeactivated = jest.fn()
    const wrapper = mount(
      <Dropdown
        onActivate={onActivate}
        onActivated={onActivated}
        onDeactivate={onDeactivate}
        onDeactivated={onDeactivated}
      >
        <Dropdown.Trigger />
        <Dropdown.Content />
      </Dropdown>
    )

    const trigger = wrapper.find('.btn.dropdown-toggle')

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalled()
    expect(onActivated).toHaveBeenCalled()
    expect(onDeactivate).toHaveBeenCalledTimes(0) // not yet!
    expect(onDeactivated).toHaveBeenCalledTimes(0) // not yet!

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1) // only once
    expect(onActivated).toHaveBeenCalledTimes(1) // only once
    expect(onDeactivate).toHaveBeenCalled()
    expect(onDeactivated).toHaveBeenCalled()
  })

  test('should use passed active prop instead of internal active', () => {
    const wrapper = mount(
      <Dropdown active>
        <Dropdown.Trigger />
        <Dropdown.Content />
      </Dropdown>
    )

    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('false')

    // Should still be able to use default activate/deactivate hooks
    wrapper.find('.btn.dropdown-toggle').simulate('click')

    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('true')

    // Passing active at any point should again override internal state
    wrapper.setProps({ active: true })

    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('false')
  })

  test('should use passed activate and deactivate functions', () => {
    const activate = jest.fn()
    const deactivate = jest.fn()
    const wrapper = mount(
      <Dropdown active={false} activate={activate} deactivate={deactivate}>
        <Dropdown.Trigger />
        <Dropdown.Content />
      </Dropdown>
    )

    const trigger = wrapper.find('.btn.dropdown-toggle')

    trigger.simulate('click')

    // We have overridden the activate event so, dropdown should still be closed
    // and clicking trigger should call activate again

    expect(activate).toHaveBeenCalledTimes(1)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('true')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('true')

    // Mock controlled component being passed true, click should now ONLY call deactivate
    wrapper.setProps({ active: true })
    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(1)
    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('false')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(2)
    expect(wrapper.find('.dropdown-content').prop('aria-hidden')).toEqual('false')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
it('renders correctly', () => {
  const tree = renderer
    .create(
      <Dropdown>
        <Dropdown.Trigger>Trigger</Dropdown.Trigger>
        <Dropdown.Content>
          <Dropdown.Item>Item 1</Dropdown.Item>
          <Dropdown.Item>Item 2</Dropdown.Item>
        </Dropdown.Content>
      </Dropdown>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
