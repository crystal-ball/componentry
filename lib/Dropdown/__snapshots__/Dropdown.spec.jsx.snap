// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="dropdown"
  data-test="dropdown-container"
  onMouseEnter={undefined}
  onMouseLeave={undefined}
>
  <button
    aria-controls={null}
    aria-describedby={null}
    aria-expanded="false"
    aria-haspopup="true"
    aria-hidden={null}
    aria-labelledby={null}
    className="btn dropdown-toggle btn-primary"
    data-test="dropdown-toggle"
    id="guid"
    onClick={[Function]}
    role={undefined}
    type="button"
  >
    Trigger
  </button>
  <div
    aria-controls={null}
    aria-describedby={null}
    aria-expanded={null}
    aria-haspopup={null}
    aria-hidden="true"
    aria-labelledby="guid"
    className="dropdown-content"
    data-test="dropdown-content"
    id={null}
    role={undefined}
  >
    <button
      className="btn dropdown-item btn-primary"
      type="button"
    >
      Item 1
    </button>
    <button
      className="btn dropdown-item btn-primary"
      type="button"
    >
      Item 2
    </button>
  </div>
</div>
`;
